### PR TITLE
Retrieve the local_process_rank from the LOCAL_RANK environment variable.

### PR DESCRIPTION
**Pull request by anw90**
This PR aims to resolve the problem of launching a distributed job using torchrun.

As described in this [issue](https://github.com/pytorch/xla/issues/5771), there is an issue when using torchrun. After testing the latest code of torch_xla, the problem has undergone some changes.

In the latest code of torch_xla, the [`local_process_rank`](https://github.com/pytorch/xla/blob/2e6e183e0724818f137c8135b34ef273dea33318/torch_xla/csrc/runtime/pjrt_registry.cc#L90C1-L91C1) is always 0, causing all processes to run on GPU:0 when using torchrun.

With torchrun, the value of `local_process_rank` should be obtained from `LOCAL_RANK`.

@vanbasten23 @JackCaoG

### CODE REVIEW

File: torch_xla/csrc/runtime/pjrt_registry.cc [hunk 1]
```diff
@@ -87,7 +87,7 @@ InitializePjRt(const std::string& device_type) {
              device_type == "ROCM") {
     TF_VLOG(1) << "Initializing PjRt GPU client...";
     bool async = sys_util::GetEnvBool(env::kEnvPjrtAsyncGpuClient, true);
-    int local_process_rank = sys_util::GetEnvInt(env::kEnvPjRtLocalRank, 0);
+    int local_process_rank = sys_util::GetEnvInt("LOCAL_RANK", 0);
```
Comment (will-cromar) [inline torch_xla/csrc/runtime/pjrt_registry.cc:90]: `LOCAL_RANK` isn't always set in `xmp.spawn` due to compatiblity issues. What about `sys_util::GetEnvInt(env::kEnvPjRtLocalRank, sys_util::GetEnvInt(env::"LOCAL_RANK", 0););`? The same thing should work for `WORLD_SIZE` with `PjRtLocalWorldSize`.

**Comment by vanbasten23** [comment]
Here is what I see:
```
# TF_CPP_MIN_LOG_LEVEL=0 TF_CPP_VMODULE=pjrt_registry=5  PJRT_DEVICE=CUDA torchrun --nnodes=1 --node_rank=0 --nproc_per_node=4 pytorch/xla/test/test_train_mp_imagenet.py --fake_data --pjrt_distributed --batch_size=16 --num_epochs=1 --model=resnet18

2024-02-26 23:12:45.275873: I torch_xla/csrc/runtime/pjrt_registry.cc:134] Getting StreamExecutorGpuClient for node_id=1, num_nodes=4, spmd case=0, PJRT_LOCAL_PROCESS_RANK=1, RANK=1, LOCAL_WORLD_SIZE=4, WORLD_SIZE=4
2024-02-26 23:12:45.275910: I external/tsl/tsl/platform/default/grpc_credentials.cc:30] gRPC insecure client credentials are used.
2024-02-26 23:12:45.275952: I torch_xla/csrc/runtime/pjrt_registry.cc:127] Initializing PjRt GPU client...
2024-02-26 23:12:45.276015: I torch_xla/csrc/runtime/pjrt_registry.cc:134] Getting StreamExecutorGpuClient for node_id=2, num_nodes=4, spmd case=0, PJRT_LOCAL_PROCESS_RANK=2, RANK=2, LOCAL_WORLD_SIZE=4, WORLD_SIZE=4
2024-02-26 23:12:45.276037: I external/tsl/tsl/platform/default/grpc_credentials.cc:30] gRPC insecure client credentials are used.
2024-02-26 23:12:45.296517: I torch_xla/csrc/runtime/pjrt_registry.cc:127] Initializing PjRt GPU client...
2024-02-26 23:12:45.296588: I torch_xla/csrc/runtime/pjrt_registry.cc:134] Getting StreamExecutorGpuClient for node_id=0, num_nodes=4, spmd case=0, PJRT_LOCAL_PROCESS_RANK=0, RANK=0, LOCAL_WORLD_SIZE=4, WORLD_SIZE=4
2024-02-26 23:12:45.306185: I torch_xla/csrc/runtime/pjrt_registry.cc:134] Getting StreamExecutorGpuClient for node_id=3, num_nodes=4, spmd case=0, PJRT_LOCAL_PROCESS_RANK=3, RANK=3, LOCAL_WORLD_SIZE=4, WORLD_SIZE=4
```
as you can see, PJRT_LOCAL_PROCESS_RANK is not always 0.

So I went further to check the `nvidia-smi` result:
```
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1348      C   /usr/local/bin/python                     12976MiB |
|    0   N/A  N/A      1349      C   /usr/local/bin/python                       306MiB |
|    0   N/A  N/A      1350      C   /usr/local/bin/python                       306MiB |
|    0   N/A  N/A      1351      C   /usr/local/bin/python                       306MiB |
|    1   N/A  N/A      1349      C   /usr/local/bin/python                     12736MiB |
|    2   N/A  N/A      1350      C   /usr/local/bin/python                     12976MiB |
|    3   N/A  N/A      1351      C   /usr/local/bin/python                     12736MiB |
+---------------------------------------------------------------------------------------+
```
It got better than in the original issue but still it seems that all process tried to access one GPU device

File: torch_xla/csrc/runtime/pjrt_registry.cc (hunk 1 above)
Comment (anw90) [reply torch_xla/csrc/runtime/pjrt_registry.cc:90]: Okay. By the way, is it possible to align all distributed environment variables with torchrun, so that we don't need to deal with such issues?

**Comment by yitongh** [comment]
@vanbasten23 This is because dist init_process_group sets PJRT_LOCAL_PROCESS_RANK through the chain of calls: dist.register_rendezvous_handler -> rendezvous.pjrt_rendezvous_handler -> pjrt.initialize_multiprocess. However, if the user utilizes any xm-related APIs such as xm.xla_device before init_process_group, it will result in all processes being on GPU 0.

We may need to initialize the call to pjrt.initialize_multiprocess in certain places, such as in __init__.py, for the torchrun environment.

Moreover, having all processes occupy memory on GPU 0 is also a serious issue.

**Comment by anw90** [comment]
> However, if the user utilizes any xm-related APIs such as xm.xla_device before init_process_group, it will result in all processes being on GPU 0.

@yitongh @vanbasten23 For most torch-xla users, they do not initialize any process group in their code. They simply use xm.all_reduce to reduce gradients. In this case, all processes will run on GPU:0.

File: torch_xla/csrc/runtime/pjrt_registry.cc (hunk 1 above)
Comment (will-cromar) [reply torch_xla/csrc/runtime/pjrt_registry.cc:90]: This is what we did originally. I changed it because it caused some issue in HuggingFace's Trainer, which was relying on the presence of `LOCAL_RANK` for some unrelated behavior: https://github.com/pytorch/xla/pull/4646

Not sure if this issue is still relevant. I would prefer to use `LOCAL_RANK` and `LOCAL_WORLD_SIZE` in principle though.

Comment (vanbasten23) [inline torch_xla/csrc/runtime/pjrt_registry.cc:90]: How do you feel about using a default value of -1 and then assert local_process_rank to be a non-negative, since we expect either kEnvPjRtLocalRank or LOCAL_RANK to be set and don't want to use a default value?

**Comment by vanbasten23** [comment]
SGTM. Is it possible to add a test?

File: torch_xla/csrc/runtime/pjrt_registry.cc (hunk 1 above)
Comment (will-cromar) [reply torch_xla/csrc/runtime/pjrt_registry.cc:90]: I don't think this would work if you were prototyping in a notebook or REPL. Neither of these will be set automatically. `0` is a reasonable default value IMO.

**Comment by anw90** [comment]
Problem solved in https://github.com/pytorch/xla/pull/6289

**Comment by vanbasten23** [comment]
> Problem solved in #6289

@anw90 Do you mean #6289 resolved this issue https://github.com/pytorch/xla/issues/5771?

**Comment by anw90** [comment]
@vanbasten23 The issue https://github.com/pytorch/xla/issues/5771 is partly resolved. I tested it on the latest PyTorch/XLA with the following command:

```
PJRT_DEVICE=GPU torchrun --nnodes 1 --nproc_per_node 4 pytorch/xla/test/test_train_mp_imagenet.py  --fake_data --pjrt_distributed --batch_size=32 --num_epochs=1
```

Here is the nvidia-smi information:

```code
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.161.07             Driver Version: 535.161.07   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10                     On  | 00000000:00:07.0 Off |                    0 |
|  0%   38C    P0              59W / 150W |  18172MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A10                     On  | 00000000:00:08.0 Off |                    0 |
|  0%   38C    P0              59W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA A10                     On  | 00000000:00:09.0 Off |                    0 |
|  0%   41C    P0              61W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA A10                     On  | 00000000:00:0A.0 Off |                    0 |
|  0%   39C    P0              58W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   4043963      C   /usr/bin/python                           17478MiB |
|    0   N/A  N/A   4043964      C   /usr/bin/python                             226MiB |
|    0   N/A  N/A   4043965      C   /usr/bin/python                             226MiB |
|    0   N/A  N/A   4043966      C   /usr/bin/python                             226MiB |
|    1   N/A  N/A   4043964      C   /usr/bin/python                           17478MiB |
|    2   N/A  N/A   4043965      C   /usr/bin/python                           17478MiB |
|    3   N/A  N/A   4043966      C   /usr/bin/python                           17478MiB |
+---------------------------------------------------------------------------------------+
```

We can see that all processes allocate a context on `GPU:0`. So as I mentioned before, I make some changes in pytorch(https://github.com/AlibabaPAI/pytorch/commit/2688630fa9f6afee6cbcd8c90b69ae1faa528f90) to fix this issue. And here is the final nvidia-smi information:

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.161.07             Driver Version: 535.161.07   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10                     On  | 00000000:00:07.0 Off |                    0 |
|  0%   38C    P0              59W / 150W |  18172MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A10                     On  | 00000000:00:08.0 Off |                    0 |
|  0%   38C    P0              59W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA A10                     On  | 00000000:00:09.0 Off |                    0 |
|  0%   41C    P0              61W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA A10                     On  | 00000000:00:0A.0 Off |                    0 |
|  0%   39C    P0              58W / 150W |  17483MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   4043963      C   /usr/bin/python                           17478MiB |
|    1   N/A  N/A   4043964      C   /usr/bin/python                           17478MiB |
|    2   N/A  N/A   4043965      C   /usr/bin/python                           17478MiB |
|    3   N/A  N/A   4043966      C   /usr/bin/python                           17478MiB |
+---------------------------------------------------------------------------------------+
```

Everything seems ok now.

**Comment by vanbasten23** [comment]
Thanks for confirming!
